Enable GIC v2 SMP support and adding NXP ls1046a RDB board support

JiafeiPan opened the pull request:
This patchset includes the following update:
1. Enable SMP support for ARM GIC v2.
2. Fix some SMP issues.
3. Add NXP ls1046a (four A72 Cores) RDB board support.

Please help to review, thanks.

JiafeiPan committed:
board: arm64: add nxp ls1046a RDB board support

Enabled ARM A72 Core, MMU, GIC and UART.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
arch: arm64: add 40 bits physical and virtual address

Add support 40 bits physical and virtual address space.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

JiafeiPan committed:
soc: arm64: add NXP ls1046a support

NXP ls1046a is SoC includes the following functions and features:
1. Four 32/64-bit Arm v8 A72 CPUs
2. Cache coherent interconnect (CCI-400)
3. One 32/64-bit DDR4 SDRAM memory controller with ECC
4. Peripheral interfaces include DPAA network, PCIe, USB, SPI,
   GPIO, UART, SDHC, IFC, LPUART ...

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

carlocaione commented on drivers/interrupt_controller/intc_gic.c:
```diff
@@ -210,3 +226,11 @@ int arm_gic_init(const struct device *unused)
+#ifdef CONFIG_SMP
+void arm_gic_secondary_init(void)
+{
+	/* Init CPU interface registers for each secondary Cores*/
```
nit: `each secondary core`

JiafeiPan replied:
Will update, thanks.

carlocaione commented on drivers/interrupt_controller/intc_gic.c:
```diff
@@ -93,6 +93,22 @@ void arm_gic_eoi(unsigned int irq)
+		GICD_SGIR_CPULIST(target_list & GICD_SGIR_CPULIST_MASK) |
+		sgi_id;
+
+	__DSB();
+	sys_write32(sgi_val, GICD_SGIR);
+	__ISB();
```
`dsb()` and `isb()`

JiafeiPan replied:
will update, thanks.

JiafeiPan replied:
Hi, @carlocaione ,

I replace __DSB with dsb() and __ISB() with isb(), there is no issue for ARM64, but there will be build warning on ARM platform, for example for board: mercury_xu:

drivers/interrupt_controller/intc_gic.c:107:2: warning: implicit declaration of function 'dsb' [-Wimplicit-function-declaration]
  107 |  dsb();
      |  ^~~
drivers/interrupt_controller/intc_gic.c:109:2: warning: implicit declaration of function 'isb' [-Wimplicit-function-declaration]
  109 |  isb();
      |  ^~~

but there is no such issue when we use __DSB() and __ISB(), so I would like to rollback this modification, what's your comments? thanks.

Jiafei.

carlocaione commented on soc/arm64/nxp_layerscape/ls1046a/mmu_regions.c:
```diff
@@ -0,0 +1,33 @@
+	MMU_REGION_FLAT_ENTRY("GIC",
+			      DT_REG_ADDR_BY_IDX(DT_INST(0, arm_gic), 0),
+			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 0),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
```
Use one of `MT_P_RW_U_RW` instead of simply `MT_RW`.

Also any reason to not use `MT_DEFAULT_SECURE_STATE`?

JiafeiPan replied:
No special reason, will update it, thanks.

carlocaione commented on soc/arm64/nxp_layerscape/ls1046a/mmu_regions.c:
```diff
@@ -0,0 +1,33 @@
+			      DT_REG_SIZE_BY_IDX(DT_INST(0, arm_gic), 1),
+			      MT_DEVICE_nGnRnE | MT_RW | MT_NS),
+
+	MMU_REGION_FLAT_ENTRY("DRAM",
+			      CONFIG_SRAM_BASE_ADDRESS,
+			      CONFIG_SRAM_SIZE * 1024,
+			      MT_NORMAL | MT_RW | MT_NS),
```
Under normal circumstances there would be no need to map the whole SRAM. Any reason why you are mapping the whole space?

JiafeiPan replied:
I only map the SRAM used by Zephyr (1M bytes), any other concerns? thanks.

Jiafei.

carlocaione replied:
In general there would be no reason to map any SRAM region at all since all the relevant regions are already mapped when MMU is initialized.

JiafeiPan replied:
I found the following MMU mapping table for ARM64, so these three regisons will cover the whole SRAM space defined by CONFIG_SRAM_SIZE and CONFIG_SRAM_BASE_ADDRESS? thanks.

 620 static const struct arm_mmu_flat_range mmu_zephyr_ranges[] = {
 621
 622         /* Mark the zephyr execution regions (data, bss, noinit, etc.)
 623          * cacheable, read-write
 624          * Note: read-write region is marked execute-never internally
 625          */
 626         { .name  = "zephyr_data",
 627           .start = _image_ram_start,
 628           .end   = _image_ram_end,
 629           .attrs = MT_NORMAL | MT_P_RW_U_NA | MT_DEFAULT_SECURE_STATE },
 630
 631         /* Mark text segment cacheable,read only and executable */
 632         { .name  = "zephyr_code",
 633           .start = _image_text_start,
 634           .end   = _image_text_end,
 635           .attrs = MT_NORMAL | MT_P_RX_U_RX | MT_DEFAULT_SECURE_STATE },
 636
 637         /* Mark rodata segment cacheable, read only and execute-never */
 638         { .name  = "zephyr_rodata",
 639           .start = _image_rodata_start,
 640           .end   = _image_rodata_end,
 641           .attrs = MT_NORMAL | MT_P_RO_U_RO | MT_DEFAULT_SECURE_STATE },
 642 };

carlocaione replied:
No, the three regions are the minimum set necessary to have Zephyr running and these will not cover the whole SRAM space. The remaining SRAM space is mapped at run-time only when needed using usually `arch_mem_map()` or `arch_mem_domain_partition_add()` but that's done by the kernel and not something the user should be concerned about.

The point is that there should be very few reasons for you to map statically the whole SRAM at boot since the code is smart enough to dynamically map the SRAM regions only when needed. It is indeed so rare that someone wants to map the whole SRAM at boot that a special kconfig for that exists, see `CONFIG_ARCH_MAPS_ALL_RAM`.

JiafeiPan replied:
I see, will remove sram mapping, thanks.

carlocaione commented on soc/arm64/nxp_layerscape/ls1046a/soc.c:
```diff
@@ -0,0 +1,35 @@
+
+	ARG_UNUSED(arg);
+
+	key = irq_lock();
+
+	/*todo: add soc init code here*/
+
+	irq_unlock(key);
```
so, we can remove the lock as well given there is no code at the moment?

JiafeiPan replied:
Yes, currently we can remove it. thanks.

Jiafei.

carlocaione commented on soc/arm64/nxp_layerscape/ls1046a/soc.h:
```diff
@@ -0,0 +1,19 @@
+#include <devicetree.h>
+#endif /* _ASMLANGUAGE */
+
+#define UART_REG_ADDR_INTERVAL 1
```
is this used anywhere?

JiafeiPan replied:
Yes, it is used by uart driver: drivers/serial/uart_ns16550.c

Jiafei.

carlocaione commented on boards/arm64/nxp_ls1046ardb/nxp_ls1046ardb_defconfig:
```diff
@@ -0,0 +1,25 @@
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=1800000000
+
+# Zephyr Kernel Configuration
+CONFIG_SRAM_BASE_ADDRESS=0xc0000000
+CONFIG_SRAM_SIZE=1024
```
This should be get directly from the DT, no need to have this here.

JiafeiPan replied:
It is a good idea, but I found CONFIG_SRAM_SIZE is also used in include/arch/arm64/scripts/linker.ld

 48 #define RAM_SIZE (CONFIG_SRAM_SIZE * 1K)
 49 #define RAM_ADDR CONFIG_SRAM_BASE_ADDRESS

so any idea how we can assign the value to these macro by getting informaton from DT? thanks.

Jiafei.

carlocaione replied:
The default value for `CONFIG_SRAM_SIZE` (and also `SRAM_BASE_ADDRESS`) is taken directly from the DT, see:

https://github.com/zephyrproject-rtos/zephyr/blob/8cae4661003525102be36767d97219b00b3697f3/arch/Kconfig#L161-L163

As you can see the default value is taken directly from the DT using the `zephyr,sram` handle.

https://github.com/zephyrproject-rtos/zephyr/blob/8cae4661003525102be36767d97219b00b3697f3/arch/Kconfig#L159

JiafeiPan replied:
Fine, will update the patch, thanks.

carlocaione commented on boards/arm64/nxp_ls1046ardb/nxp_ls1046ardb_smp_defconfig:
```diff
@@ -0,0 +1,35 @@
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=1800000000
+
+# Zephyr Kernel Configuration
+CONFIG_SRAM_BASE_ADDRESS=0xc0000000
+CONFIG_SRAM_SIZE=1024
```
same as before

JiafeiPan replied:
Removed, thanks.

JiafeiPan replied:
All done, thanks for yoru comments.

Jiafei.

JiafeiPan committed:
CODEOWNERS: add myself as Layerscape code-owner

Add myself to be NXP Layserscape SoC and LS1046A board
code-owner.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>

carlocaione commented on boards/arm64/qemu_cortex_a53/qemu_cortex_a53_smp_defconfig:
```diff
@@ -25,3 +25,4 @@ CONFIG_MP_NUM_CPUS=2
 CONFIG_CACHE_MANAGEMENT=y
 CONFIG_TIMEOUT_64BIT=y
 CONFIG_ARMV8_A_NS=y
+CONFIG_MAX_XLAT_TABLES=12
```
It's a bit odd that you need to increase the number of xlat tables. I suggest you to investigate why you actually need this, I suspect this is hiding a different problem somewhere.

JiafeiPan replied:
Yes, it is odd, I can't find out why my pathces impact qemu_cotex_a53_smp board. most likely I modify GICV2 driver, but it seems qemu_cotex_a53_smp board is using GICv3, right?

carlocaione replied:
I think the problem is that setting the VA/PA size to 40-bits there are now more tables to deal with when copying / duplicating, if this is correct then it's fine to increment the number of XLAT tables here. I'll try to investigate, meanwhile feel free to increment this number to the minimum needed to pass the tests.

JiafeiPan replied:
I removed SMP related patches but keep VA/PA 40bit patch, now all checks have passed, so the issue should be not introduced by VA/PA 40bit patches, I would like to submit non-SMP patches firstly, and then submit SMP patches in another pull request, thanks.

JiafeiPan commented:
Updated the pull request:
1. Update CODEOWNERS to add myself as Layerscape code-owner.
2. Fixed Kconfig item SOC_PART_NUMBER_LS1046A not defined building warning.
2. Fix build warning for arm boards caused by using dsb() and isb() in GIC driver "drivers/interrupt_controller/intc_gic.c".
3. Removed SMP related patches, I would like to submit SMP support in another pull request

Now all checks have passed, please review again, thanks.

carlocaione commented on boards/arm64/nxp_ls1046ardb/doc/index.rst:
```diff
@@ -0,0 +1,128 @@
+
+Or for SMP:
+
+	``boards/arm64/nxp_ls1046ardb/nxp_ls1046ardb_smp_defconfig``
```
Do you want to remove this if you are going to push SMP at a later time?

JiafeiPan replied:
Done, I removed SMP related document, please review again, thanks.

JiafeiPan commented:
Currently there is a build failure in CI, I believe this failure has no relation with my patch, I only updated a document, before that all test have passed, so how to retriger CI to build agian or whether patches can be merged? thanks.


INFO    - 301/301 mps2_an385                tests/net/socket/select/net.socket.select.preempt  FAILED Failed (qemu 2.238s)
--
  | INFO    - /workdir/zephyr/twister-out/mps2_an385/tests/net/socket/select/net.socket.select.preempt/handler.log
  | ERROR   - *** Booting Zephyr OS build zephyr-v2.5.0-2067-g2edbe44d82c5  ***
  | Running test suite socket_select
  | ===================================================================
  | START - test_fd_set
  | PASS - test_fd_set in 0.2 seconds
  | ===================================================================
  | START - test_select
  |  
  | Assertion failed at WEST_TOPDIR/zephyr/tests/net/socket/select/src/main.c:101: test_select: (tstamp <= FUZZ is false)
  |  
  | FAIL - test_select in 0.103 seconds
  | ===================================================================
  | Test suite socket_select failed.
  | ===================================================================
  | PROJECT EXECUTION FAILED

carlocaione commented:
> Currently there is a build failure in CI, I believe this failure has no relation with my patch, I only updated a document, before that all test have passed, so how to retriger CI to build agian or whether patches can be merged?

You can close and reopen the PR or just force-push again after rebasing on master.

JiafeiPan commented:
> > Currently there is a build failure in CI, I believe this failure has no relation with my patch, I only updated a document, before that all test have passed, so how to retriger CI to build agian or whether patches can be merged?
> 
> You can close and reopen the PR or just force-push again after rebasing on master.

Rebased, thanks.
